src/react/hooks/useMutation.ts

```diff
@@ -119,27 +130,49 @@ export declare namespace useMutation {
 
   export type ResultTuple<
     TData,
-    TVariables,
+    TVariables extends OperationVariables,
     TContext = DefaultContext,
     TCache extends ApolloCache = ApolloCache,
   > = [
     mutate: (
-      options?: MutationFunctionOptions<TData, TVariables, TContext, TCache>
+      ...[options]: {} extends TVariables ?
+        [
+          options?: MutationFunctionOptions<
+            TData,
+            TVariables,
+            TContext,
+            TCache
+          > & {
+            /** {@inheritDoc @apollo/client!MutationOptionsDocumentation#variables:member} */
+            variables?: TVariables;
+          },
+        ]
+      : [
+          options: MutationFunctionOptions<
+            TData,
+            TVariables,
+            TContext,
+            TCache
+          > & {
+            /** {@inheritDoc @apollo/client!MutationOptionsDocumentation#variables:member} */
+            variables: TVariables;
+          },
+        ]
     ) => Promise<MutateResult<MaybeMasked<TData>>>,
     result: Result<TData>,
   ];
 
```

Comment:
Let's add a `MutationFunction` type here. This might be useful for reuse in userland.

```suggestion
  type MutationFunction<
    TData,
    TVariables extends OperationVariables,
    TContext = DefaultContext,
    TCache extends ApolloCache = ApolloCache,
  > = (
    ...[options]: {} extends TVariables ?
      [
        options?: MutationFunctionOptions<
          TData,
          TVariables,
          TContext,
          TCache
        > & {
          /** {@inheritDoc @apollo/client!MutationOptionsDocumentation#variables:member} */
          variables?: TVariables;
        },
      ]
    : [
        options: MutationFunctionOptions<
          TData,
          TVariables,
          TContext,
          TCache
        > & {
          /** {@inheritDoc @apollo/client!MutationOptionsDocumentation#variables:member} */
          variables: TVariables;
        },
      ]
  ) => Promise<MutateResult<MaybeMasked<TData>>>;

  export type ResultTuple<
    TData,
    TVariables extends OperationVariables,
    TContext = DefaultContext,
    TCache extends ApolloCache = ApolloCache,
  > = [
    mutate: MutationFunction<TData, TVariables, TContext, TCache>,
    result: Result<TData>,
  ];

```